### PR TITLE
Modified EfficientDet

### PR DESCRIPTION
**Pull request by shivam-raj**
Heavier model based on efficientdet using a new FPN called QuadFPN (Basically a BiFPN and inverted BiFPN connected in parallel and a final FPN fusing features from parallel paths. For the architecture and checkpoints kindly refer to README of this branch. 

I have been experimenting on parallel fusion techniques and thus extracting more from backbone that one FPN can. 
A fully parallel FPN demo ,architecture and checkpoints are available on the master branch of my fork of this repo.
(Tho accuracy remains almost same but it is very inefficient and doesn't scale well. QuadFPN proves to be a better choice).

Another variant that even performed better was using unweighted QuadFPN (Unweighted parallel FPN connections and weighted connection for contribution of parallel  FPNs).Set fpn_quad_method='fastattn' and fpn_weight_method='sum' in hparams_config.py to use this. This provides more stable learning and better accuracy. Unfortunately my TFRC quota finished before I could do any further training.

### CODE REVIEW

**Comment by shivam-raj** [comment]
I am sorry but my fork is outdated by 2 months. I've created this PR just for review. My checkpoints don't really work with the latest code. Maybe retraining with latest code is required idk

File: efficientdet/efficientdet_arch.py [hunk 1]
```diff
@@ -505,34 +476,43 @@ def build_feature_network(features, config):
...
   p.nodes = [
-      {'feat_level': 6, 'inputs_offsets': [3, 4]},
-      {'feat_level': 5, 'inputs_offsets': [2, 5]},
-      {'feat_level': 4, 'inputs_offsets': [1, 6]},
-      {'feat_level': 3, 'inputs_offsets': [0, 7]},
-      {'feat_level': 4, 'inputs_offsets': [1, 7, 8]},
-      {'feat_level': 5, 'inputs_offsets': [2, 6, 9]},
-      {'feat_level': 6, 'inputs_offsets': [3, 5, 10]},
-      {'feat_level': 7, 'inputs_offsets': [4, 11]},
+      {'feat_level': 6, 'inputs_offsets': [3, 4] , 'weight_method': p.weight_method},
```
Comment (mingxingtan) [inline efficientdet/efficientdet_arch.py:485]: Is this the main change? So now each layer has four paths: 

(top down + bottom up)  + (bottom_up + top_down)

Any reason to have  two different weight_methods?

File: efficientdet/hparams_config.py [hunk 2]
```diff
@@ -275,7 +223,7 @@ def default_detection_configs():
             backbone_name='efficientnet-b1',
             image_size=640,
             fpn_num_filters=88,
-            fpn_cell_repeats=4,
+            fpn_cell_repeats=5,
```
Comment (mingxingtan) [inline efficientdet/hparams_config.py:226]: Each original BiFPN layer has 2 paths. With fpn_cell_repeats=4, it have 2 * 4 =8 paths in total.
Here you change each layer to have 4 paths. With fpn_cell_repeats=5, you will have 4*5=20 paths in total. 
Is my understanding correct?

**Comment by mingxingtan** [comment]
> I am sorry but my fork is outdated by 2 months. I've created this PR just for review. My checkpoints don't really work with the latest code. Maybe retraining with latest code is required idk

No worries.

File: efficientdet/efficientdet_arch.py (hunk 1 above)
Comment (shivam-raj) [reply efficientdet/efficientdet_arch.py:485]: image

Yea, basically creating this structure. At first I used fast normalized fusion for the entire FPN , but then I experimented a bit and noticed the network performs even better when you use unweighted connections in the two parallel FPNs and weighted connection for fusion of the two. This gave higher accuracy and better stability. I didn't get to experimenting this variant with heavier models though. So still requires training and testing.

File: efficientdet/hparams_config.py (hunk 2 above)
Comment (shivam-raj) [reply efficientdet/hparams_config.py:226]: Yes absolutely right ! It's scales very differently and not as good as EfficientDets but is able to extract more features and hence better accuracy. I tried scaling BiFPN (increases fpn cell repeats) to match the same number of paths in QuadFPN but couldn't increases performance more than 0.1 - 0.2% AP than one reported in your paper , so thought maybe try this out further.

**Comment by shivam-raj** [comment]
@mingxingtan Hey any leads ?
I've provided the checkpoints in my README
Can you just tell me I should pursue this further or not ? 😬

**Pull request closed without merging**